### PR TITLE
net: icmpv4: Simplify the flow at net_icmpv4_get_hdr()

### PR DESCRIPTION
**Pull request by ozhuraki**
Minor simplification here, no need to trigger 3 reads here,
net_frag_read() will do the same inside itself.

Signed-off-by: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>

### CODE REVIEW

**Review by jukkar — CHANGES_REQUESTED** [review]
The refactoring is missing the fast path here. Typically in 99% of the cases the net_icmp_header_fits() returns true in which case we do not need to do the reading. Thus I do not see much point in this refactoring.

**Comment by tbursztyka** [comment]
Actually this patch makes totally sense.

net_pkt_icmp_data() isn't faster if the header fits into the frag already, here is the call trace:

```
net_pkt_icmp_data -> net_pkt_get_pos -> net_frag_skip -> net_frag_read
```

(with tests on frag validity, length..)

Where here, it would just do give:

```
net_frag_read directly.
```

And if it does not fit, it will anyway go to net_frag_read.

**Review by tbursztyka — APPROVED** [review]

File: subsys/net/ip/icmpv4.c [hunk 1]
```diff
@@ -55,26 +55,12 @@ struct net_icmp_hdr *net_icmpv4_set_hdr(struct net_pkt *pkt,
...
-	frag = net_frag_read_u8(pkt->frags, net_pkt_ip_hdr_len(pkt), &pos,
-				&hdr->type);
-	frag = net_frag_read_u8(frag, pos, &pos, &hdr->code);
-	frag = net_frag_read(frag, pos, &pos, sizeof(hdr->chksum),
-			     (u8_t *)&hdr->chksum);
-	if (!frag) {
-		NET_ASSERT(frag);
-		return NULL;
-	}
+	u16_t pos = 0;
```
Comment (tbursztyka) [inline subsys/net/ip/icmpv4.c]: you need to start at pos net_pkt_ip_hdr_len(pkt)

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Comment by tbursztyka** [comment]
@ozhuraki you could do the same for net_icmpv4_set_hdr() btw

**Review by tbursztyka — APPROVED** [review]

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by pfalcon** [comment]
@jukkar : So, what happened with you claim about fast path handling in this PR?

**Comment by pfalcon** [comment]
> My claim was for earlier version of the PR where it called more heavier net_pkt_frag*() function always. Calling net_pkt_linearize() is quite elegant now.

And yet the fast path you were talking about is now gone, as net_pkt_linearize() uses memcpy().

The issue here is that we have the same pattern (if a header fits within frag, return ptr, otherwise make copy) in many other places. So, this PR sets an important precedent of getting rid of it. I think it's especially important to draw attention to that in the light of what @tbursztyka was talking today at the forum meeting (code cleanup). Myself, I'm split in feelings re: this "fast path" pattern. On one hand, I love optimizations. On the other hand, this adds branches to bunch of functions, and I'm sure that non-fastpath case isn't covered by the tests we have.

**Pull request merged**